Make Formats.parsing method non-private

PeterPerhac opened the pull request:
For users developing their own `Formatter`s wanting to re-use the logic in the `parsing` method, please make this method non-private.

# Pull Request Checklist

* [ ] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [ ] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you [squashed your commits]? (Optional, but makes merge messages nicer.)
* [ ] Have you added copyright headers to new files?
* [ ] Have you checked that both Scala and Java APIs are updated?
* [ ] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

# Helpful things

## Fixes

Fixes #xxxx

## Purpose

What does this PR do?

## Background Context

Why did you take this approach?

## References

Are there any relevant issues / PRs / mailing lists discussions?

typesafehub-validator commented:
Hi @PeterPerhac,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

wsargent commented:
Hi @PeterPerhac can you add some documentation and tests to show when and where you'd use this method in an application?

PeterPerhac commented:
Hello @wsargent I put together an "ACME" sample of how I would implement binding of a String to a property of type java.net.URL. The binding `of[URL]` requires a Formatter for the URL type.
https://github.com/PeterPerhac/fuel-meter/blob/master/app/models/forms/AcmeForm.scala

Here I wanted to make use of the `parsing` function but, as it is marked private, I copy-pasted it into my project: (note that i was able to simply fully qualify/import the stringFormat function)
https://github.com/PeterPerhac/fuel-meter/blob/master/app/models/forms/package.scala

here is a working example of this form:
http://fuel-meter.herokuapp.com/acme/capture

marcospereira commented:
Hey @PeterPerhac,

Could you update your PR to add these examples at our documentation? The best place is probably at [ScalaForms.md](https://github.com/playframework/playframework/blob/master/documentation/manual/working/scalaGuide/main/forms/ScalaForms.md). There is a lack of documentation for this specific feature, so it would be awesome if you could add it to this PR.

Also, since the method will be public, we should have some tests/specs. The best place to add the tests is [framework/src/play/src/test/scala/play/api/data/format/FormatSpec.scala](https://github.com/playframework/playframework/blob/master/framework/src/play/src/test/scala/play/api/data/format/FormatSpec.scala).

PeterPerhac commented:
Hi I added tests and updated documentation but for some reason Travis keeps rejecting my changes with (what seems to me like completely unrelated) error in AkkaHttpWebSocketSpec

there's quite a line of red builds though... can you please advise what i could do to get the build to pass? I don't see how my changes to documentation, which I have tested buy publishing and running locally, can cause failure of AkkaHttpWebSocketSpec test

marcospereira commented:
@PeterPerhac, thank you.

About the built, it could be a glitch. I've restarted it to see if it works as expected.

marcospereira commented:
@PeterPerhac now the build failed with a good reason:

https://travis-ci.org/playframework/playframework/builds/188572872#L2487

marcospereira commented on documentation/manual/working/scalaGuide/main/forms/code/ScalaForms.scala:
```diff
@@ -429,6 +431,16 @@ class Application extends Controller with I18nSupport {
+  //#userForm-default
+  Form(
+    mapping(
+      "name" -> default(text, "Bob")
```
There is a missing "," here. That is why the build failed.

PeterPerhac replied:
See, i should not have touched it. It was someone else sticking code directly into markdown and I moved it into the Scala file and referenced it from markdown. thanks for spotting that. i thought i would make minor cosmetic adjustment while i am at it... 🤦‍♂️

marcospereira replied:
It was good that you moved it to Scala code so we could catch the error. 👍 

Ideally, all the examples (excluding the examples at migrations guides) would be extracted like this so the could would be compiled and, in some cases, even tested.

marcospereira commented:
Thanks, @PeterPerhac.
